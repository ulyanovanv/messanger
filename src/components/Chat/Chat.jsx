import React, { Component } from 'react';
import shortid from 'shortid';
import PropTypes from 'prop-types';

import images from '../../helpers/images.js';

export default class Chat extends Component {
  constructor(props) {
    super(props);

    this.scrollRed = React.createRef();

    this.renderMessage = this.renderMessage.bind(this);
    // this.renderAuthorImage = this.renderAuthorImage.bind(this);
    this.scrollDown = this.scrollDown.bind(this);
  }

  componentDidMount() {
    this.scrollDown();
  }

  componentDidUpdate() {
    this.scrollDown();
  }

  scrollDown() {
    if (document.getElementsByClassName('App__chats_message')[0]) {
      const chartScrollHeight = document.getElementsByClassName('App__chats_message')[0].scrollHeight;
      this.scrollRed.current.scrollTo(0, chartScrollHeight);
    }
  }

  renderAuthorImage(message, condition) {
    const myImage = <div className="App__chats_message_my-image mr-2"> me </div>;
    const userImage = (
      <img
        src={images[message.user]}
        className="App__chats_message_user-image mr-2"
        title={message.user}
        alt={message.user}
      />
    );

    return condition ? userImage : myImage;
  }

  renderMessage(message, condition) {
    const messageContent = message.message.includes('image_')
      ? (
        <img
          src={window.localStorage.getItem(message.message)}
          className="App__chats_message_loaded-image rounded"
          id={message.message}
          onLoad={this.scrollDown}
          alt={message.user}
        />
      ) : message.message;

    return (
      <div className={`p-1 rounded ${condition ? 'bg-light text-dark' : 'bg-secondary text-white'}`}>
        {messageContent}
        {' '}
        {' '}
        <span className={`App__chats_message_date ${
          condition ? 'text-secondary align-self-start' : 'text-light align-self-end'}`}
        >
          {message.date}
        </span>
        <br />
      </div>
    );
  }

  render() {
    const { messages } = this.props;
    const prepairedMessages = messages.map((message) => {
      const condition = message.user !== 'me';

      return (
        <div
          className={`App__chats_message_message-width d-flex flex-row align-content-center mb-1 ${
            condition ? 'mr-auto' : 'ml-auto'}`}
          key={shortid.generate()}
        >
          {this.renderAuthorImage(message, condition)}
          {this.renderMessage(message, condition)}
        </div>
      );
    });

    return (
      <div className="App__chats_message d-flex flex-column m-2" ref={this.scrollRed}>
        {prepairedMessages}
      </div>
    );
  }
}

Chat.propTypes = {
  messages: PropTypes.arrayOf(PropTypes.object),
};
