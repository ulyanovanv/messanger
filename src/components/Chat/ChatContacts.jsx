import React, { Fragment } from 'react';
import PropTypes from 'prop-types';

import images from '../../helpers/images.js';

export default function ChatContacts(props) {
  const { contacts } = props;

  const chatContacts = contacts.map(obj => (
    <Fragment key={obj.name}>
      <img
        src={images[obj.image]}
        className="mx-2"
        title={obj.image}
        alt={obj.image}
      />
      <h6>{obj.name}</h6>
    </Fragment>
  ));

  return (
    <div className="App__chats_contact d-flex flex-row justify-content-center align-items-end border-bottom py-2">
      {chatContacts}
    </div>
  );
}

ChatContacts.propTypes = {
  contacts: PropTypes.arrayOf(PropTypes.object),
};
