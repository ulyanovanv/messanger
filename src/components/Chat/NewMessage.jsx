import React, { Component } from "react"
import {inject, observer } from "mobx-react/index";
import { observable } from "mobx";

@inject('store') @observer
export default class NewMessage extends Component {
  @observable imageCounter = 0;
  @observable message = '';

  constructor(props) {
    super(props);

    this.sendMessage = this.sendMessage.bind(this);
    this.uploadImage = this.uploadImage.bind(this);
    this.handleKeyDown = this.handleKeyDown.bind(this);
  }

  handleKeyDown(e) {
    let keyCode = e.keyCode || e.which;

    if (keyCode === 13) {
      e.preventDefault();
      this.sendMessage();
    }
  }

  sendMessage() {
    if (!this.message) return;

    this.props.store.addNewMessage(this.message);
    this.message = '';
  }

  uploadImage(e) {
    e.preventDefault();
    let file = e.target.files[0];

    if (file.size > 1048576) { //1 Mb
      alert("The image size should not exceed 1Mb");
      return;
    }

    let imageKey = "image_" + this.imageCounter;
    getBase64(file).then(file => {
      localStorage[imageKey] = file;
      this.imageCounter = this.imageCounter + 1;
      this.props.store.addNewMessage(imageKey);
    });
  }

  render() {
    return (
    <div className="input-group p-2 App__chats_new-message">
      <textarea
        className="form-control"
        placeholder="Enter message text"
        aria-describedby="button-addon4"
        value={this.message}
        onChange={(event) => this.message = event.target.value}
        onKeyDown={(e) => this.handleKeyDown(e)}
      />
      <div className="input-group-append" id="button-addon4">
        <label className="btn btn-outline-secondary btn-file mb-0">
          Add image
          <input
            type="file"
            value=""
            accept="image/x-png,image/gif,image/jpeg"
            onChange={(e) => this.uploadImage(e)}
          />
        </label>
        <input
          className="btn btn-outline-success"
          type="button"
          value="Send message"
          onClick={() => this.sendMessage()}
        />
      </div>
    </div>);
  }
}

const getBase64 = (file) => {
  return new Promise((resolve,reject) => {
    const reader = new FileReader();
    reader.onload = () => resolve(reader.result);
    reader.onerror = error => reject(error);
    reader.readAsDataURL(file);
  });
};
