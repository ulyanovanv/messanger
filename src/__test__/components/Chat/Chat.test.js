import React from 'react';
import renderer from 'react-test-renderer';
import Chat from '../../../components/Chat/Chat';

test('Chat', () => {
  const messages = [
    {
      message: 'Hello, how do you do?',
      date: '28.04.2019',
      user: 'ara',
    },
    {
      message: "Nice to hear you. I am fine. Let's meet today.",
      date: '28.04.2019',
      user: 'me',
    },
  ];

  const component = renderer.create(<Chat messages={messages} />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
