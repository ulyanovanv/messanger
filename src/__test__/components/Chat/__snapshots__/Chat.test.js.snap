// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chat 1`] = `
<div
  className="App__chats_message d-flex flex-column m-2"
>
  <div
    className="App__chats_message_message-width d-flex flex-row align-content-center mb-1 mr-auto"
  >
    <img
      className="App__chats_message_user-image mr-2"
      src="test-file-stub"
      title="ara"
    />
    <div
      className="p-1 rounded bg-light text-dark"
    >
      Hello, how do you do?
       
       
      <span
        className="App__chats_message_date text-secondary align-self-start"
      >
        28.04.2019
      </span>
      <br />
    </div>
  </div>
  <div
    className="App__chats_message_message-width d-flex flex-row align-content-center mb-1 ml-auto"
  >
    <div
      className="App__chats_message_my-image mr-2"
    >
       me 
    </div>
    <div
      className="p-1 rounded bg-secondary text-white"
    >
      Nice to hear you. I am fine. Let's meet today.
       
       
      <span
        className="App__chats_message_date text-light align-self-end"
      >
        28.04.2019
      </span>
      <br />
    </div>
  </div>
</div>
`;
